File: app/routers/transactions.py
import uuid
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, or_, func
from sqlalchemy.orm import load_only

from app.database import get_db
from app.models.user import User
from app.models.transaction import Transaction
from app.middleware.auth import get_current_user
from app.services.ml.classifier import classifier, CATEGORIES

router = APIRouter(prefix="/transactions", tags=["transactions"])


def _tx_dict(t: Transaction) -> dict:
    return {
        "id": str(t.id),
        "date": str(t.date),
        "description": t.description,
        "amount": float(t.amount),
        "balance": float(t.balance) if t.balance else None,
        "type": t.type,
        "category": t.category,
        "subcategory": t.subcategory,
        "merchant": t.merchant,
        "confidence": float(t.confidence) if t.confidence else None,
        "classified_by": t.classified_by,
        "is_recurring": t.is_recurring,
        "is_subscription": t.is_subscription,
        "is_bank_charge": t.is_bank_charge,
        "is_duplicate": t.is_duplicate,
        "user_verified": t.user_verified,
        "statement_id": str(t.statement_id) if t.statement_id else None,
    }


@router.get("", response_model=dict)
async def list_transactions(
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
    search: str | None = None,
    type: str | None = None,
    category: str | None = None,
    statement_id: str | None = None,
    from_date: str | None = Query(None, alias="from"),
    to_date: str | None = Query(None, alias="to"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    conditions = [
        Transaction.user_id == current_user.id,
        Transaction.is_duplicate == False,
    ]

    if search:
        conditions.append(
            or_(
                Transaction.description.ilike(f"%{search}%"),
                Transaction.merchant.ilike(f"%{search}%"),
            )
        )
    if type:
        conditions.append(Transaction.type == type)
    if category:
        conditions.append(Transaction.category == category)
    if statement_id:
        conditions.append(Transaction.statement_id == uuid.UUID(statement_id))
    if from_date:
        from datetime import date
        conditions.append(Transaction.date >= date.fromisoformat(from_date))
    if to_date:
        from datetime import date
        conditions.append(Transaction.date <= date.fromisoformat(to_date))

    total_q = await db.execute(
        select(func.count()).select_from(Transaction).where(and_(*conditions))
    )
    total = total_q.scalar_one()

    q = await db.execute(
        select(Transaction)
        .where(and_(*conditions))
        .order_by(Transaction.date.desc())
        .offset((page - 1) * limit)
        .limit(limit)
    )
    txs = q.scalars().all()

    return {
        "success": True,
        "data": [_tx_dict(t) for t in txs],
        "pagination": {
            "page": page,
            "limit": limit,
            "total": total,
            "total_pages": (total + limit - 1) // limit,
        },
    }


@router.patch("/{transaction_id}", response_model=dict)
async def update_transaction(
    transaction_id: str,
    body: dict,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Transaction).where(
            Transaction.id == uuid.UUID(transaction_id),
            Transaction.user_id == current_user.id,
        )
    )
    tx = result.scalar_one_or_none()
    if not tx:
        raise HTTPException(status_code=404, detail="Transaction not found")

    old_category = tx.category
    if "category" in body:
        tx.category = body["category"]
        tx.user_verified = True
        tx.classified_by = "user"

        # Feed correction to ML classifier for online learning
        if old_category and old_category != body["category"]:
            try:
                classifier.partial_train([tx.description], [body["category"]])
            except Exception:
                pass

    if "subcategory" in body:
        tx.subcategory = body["subcategory"]
    if "description" in body:
        tx.description = body["description"]

    await db.commit()
    await db.refresh(tx)
    return {"success": True, "data": _tx_dict(tx)}


@router.get("/meta/categories", response_model=dict)
async def get_categories():
    return {"success": True, "data": CATEGORIES}


@router.get("/meta/uncertain", response_model=dict)
async def get_uncertain_transactions(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Transaction)
        .where(
            Transaction.user_id == current_user.id,
            Transaction.is_duplicate == False,
            Transaction.confidence < 0.6,
            Transaction.user_verified == False,
        )
        .order_by(Transaction.date.desc())
        .limit(50)
    )
    txs = result.scalars().all()
    return {"success": True, "data": [_tx_dict(t) for t in txs]}


@router.post("/train", response_model=dict)
async def train_classifier(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Retrain classifier using all user-verified transactions."""
    result = await db.execute(
        select(Transaction).where(
            Transaction.user_id == current_user.id,
            Transaction.user_verified == True,
            Transaction.category.isnot(None),
        )
    )
    txs = result.scalars().all()

    if len(txs) < 10:
        return {"success": False, "data": {"error": f"Need at least 10 verified transactions. You have {len(txs)}."}}

    texts = [t.description for t in txs]
    labels = [t.category for t in txs]
    report = classifier.train(texts, labels)
    return {"success": True, "data": report}

File: src/routes/transactions.ts
import { Router } from "express";
import { z } from "zod";
import { db } from "../db";
import { transactions, categories, userCorrections, entities } from "../db/schema";
import { eq, and, desc, gte, lte, ilike, or, sql } from "drizzle-orm";
import { requireAuth, type AuthRequest } from "../middleware/auth";
import { AppError } from "../middleware/errorHandler";

const router = Router();

// ─── GET /api/transactions ────────────────────────────────────────────────────

router.get("/", requireAuth, async (req: AuthRequest, res, next) => {
  try {
    const QuerySchema = z.object({
      page: z.coerce.number().min(1).default(1),
      limit: z.coerce.number().min(1).max(500).default(50),
      search: z.string().optional(),
      category: z.string().optional(),
      type: z.enum(["debit", "credit", "transfer"]).optional(),
      from: z.string().optional(),
      to: z.string().optional(),
      accountId: z.string().uuid().optional(),
      statementId: z.string().uuid().optional(),
      minAmount: z.coerce.number().optional(),
      maxAmount: z.coerce.number().optional(),
      isRecurring: z.coerce.boolean().optional(),
      isSubscription: z.coerce.boolean().optional(),
    });

    const query = QuerySchema.parse(req.query);
    const offset = (query.page - 1) * query.limit;

    const conditions = [
      eq(transactions.userId, req.user!.userId),
      eq(transactions.isDuplicate, false),
    ];

    if (query.search) {
      conditions.push(
        ilike(transactions.description, `%${query.search}%`)
      );
    }
    if (query.type) conditions.push(eq(transactions.type, query.type));
    if (query.from) conditions.push(gte(transactions.date, new Date(query.from)));
    if (query.to) conditions.push(lte(transactions.date, new Date(query.to)));
    if (query.accountId) conditions.push(eq(transactions.accountId, query.accountId));
    if (query.statementId) conditions.push(eq(transactions.statementId, query.statementId));
    if (query.isRecurring !== undefined) conditions.push(eq(transactions.isRecurring, query.isRecurring));
    if (query.isSubscription !== undefined) conditions.push(eq(transactions.isSubscription, query.isSubscription));

    const [rows, countResult] = await Promise.all([
      db
        .select()
        .from(transactions)
        .where(and(...conditions))
        .orderBy(desc(transactions.date))
        .limit(query.limit)
        .offset(offset),
      db
        .select({ count: sql<number>`count(*)` })
        .from(transactions)
        .where(and(...conditions)),
    ]);

    const total = Number(countResult[0]?.count || 0);

    res.json({
      success: true,
      data: rows,
      pagination: {
        page: query.page,
        limit: query.limit,
        total,
        totalPages: Math.ceil(total / query.limit),
      },
    });
  } catch (err) {
    next(err);
  }
});

// ─── PATCH /api/transactions/:id — update classification ─────────────────────

router.patch("/:id", requireAuth, async (req: AuthRequest, res, next) => {
  try {
    const UpdateSchema = z.object({
      categoryId: z.string().uuid().optional(),
      subcategory: z.string().max(100).optional(),
      merchant: z.string().max(300).optional(),
      entityId: z.string().uuid().optional(),
      isRecurring: z.boolean().optional(),
      isSubscription: z.boolean().optional(),
      applyToSimilar: z.boolean().optional().default(false),
    });

    const body = UpdateSchema.parse(req.body);

    // Ensure transaction belongs to user
    const [existing] = await db
      .select()
      .from(transactions)
      .where(
        and(
          eq(transactions.id, req.params.id),
          eq(transactions.userId, req.user!.userId)
        )
      )
      .limit(1);

    if (!existing) {
      throw new AppError(404, "Transaction not found");
    }

    // Update transaction
    const { applyToSimilar, ...updateData } = body;
    const [updated] = await db
      .update(transactions)
      .set({ ...updateData, userVerified: true })
      .where(eq(transactions.id, req.params.id))
      .returning();

    // Store correction for ML training data
    await db.insert(userCorrections).values({
      userId: req.user!.userId,
      transactionId: req.params.id,
      rawDescription: existing.rawDescription,
      predictedCategory: existing.subcategory,
      correctedCategoryId: body.categoryId || undefined,
      correctedSubcategory: body.subcategory || undefined,
      correctedEntity: undefined,
      correctedMerchant: body.merchant || undefined,
      applyToSimilar: applyToSimilar || false,
    });

    // Apply correction to similar transactions if requested
    if (applyToSimilar && body.categoryId) {
      const descNorm = existing.description.toLowerCase().slice(0, 30);
      await db
        .update(transactions)
        .set({ categoryId: body.categoryId, subcategory: body.subcategory, userVerified: true })
        .where(
          and(
            eq(transactions.userId, req.user!.userId),
            ilike(transactions.description, `%${descNorm}%`)
          )
        );
    }

    res.json({ success: true, data: updated });
  } catch (err) {
    next(err);
  }
});

// ─── GET /api/transactions/categories ────────────────────────────────────────

router.get("/meta/categories", requireAuth, async (req: AuthRequest, res, next) => {
  try {
    const cats = await db
      .select()
      .from(categories)
      .where(
        or(
          eq(categories.isSystem, true),
          eq(categories.userId, req.user!.userId)
        )
      )
      .orderBy(categories.name);

    res.json({ success: true, data: cats });
  } catch (err) {
    next(err);
  }
});

// ─── GET /api/transactions/uncertain ─────────────────────────────────────────

router.get("/meta/uncertain", requireAuth, async (req: AuthRequest, res, next) => {
  try {
    const uncertain = await db
      .select()
      .from(transactions)
      .where(
        and(
          eq(transactions.userId, req.user!.userId),
          eq(transactions.isUncertain, true),
          eq(transactions.userVerified, false)
        )
      )
      .orderBy(desc(transactions.amount))
      .limit(20);

    res.json({ success: true, data: uncertain });
  } catch (err) {
    next(err);
  }
});

export default router;
